logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java

```diff
@@ -337,63 +387,91 @@ public long write(Queueable element) throws IOException {
         // element at risk in the always-full queue state. In the later, when closing a full queue, it would be impossible
         // to write the current element.
 
-        lock.lock();
-        try {
-            if (! this.headPage.hasCapacity(data.length)) {
-                throw new IOException("data to be written is bigger than page capacity");
+        long maxSeqNum = -1;
+        int headPageCapacity = this.headPage.getPageIO().getCapacity();
+        // Our batch might span multiple pages, so we have to break it up if its too big
+
+        int persistedBatchSize = this.headPage.getPageIO().persistedByteCount(dataSize, serializedQueueables.size());
+
+        if (persistedBatchSize <= headPageCapacity) {
+            lock.lock();
+            try {
+                maxSeqNum = unsafeQueueWrite(serializedQueueables, timeoutMillis, persistedBatchSize);
+            } finally {
+                lock.unlock();
             }
+        } else {
+            // Batch is larger than a page, we can't do an atomic write
+            throw new IOException("data to be written is bigger than page capacity");
```

Comment:
@andrewvc this seems a little risky? Shouldn't we try to split the batch up across pages instead of just erroring out here? This could lead to trouble with random very large events like we've seen in the past I think.
It's obviously more logic to handle that, but maybe worth it?

Reply:
I thought about this, and in fact, my initial implementation did this (since squashed).

Unfortunately, one of the goals here is to allow for atomic writes of the batch, for things like the HTTP input. The idea is that a write can timeout and not have written anything.

I couldn't figure out a way to do that without this limitation. We could, of course, figure out some sort of way, but it seemed really complex with the way we juggle pages, and would be a major refactor. Maybe there's an easy way to do this?

Reply:
@andrewvc hmm, maybe it's not worth the trouble and we could simply write a larger page in this case?
I guess the other option would be to not be so strict about the atomic part and simply split up the list and write to multiple pages in a loop in this method.

Reply:
Spoke to @original-brownbear directly, we decided that our best option is the current code, which just makes a new page if the current one is too small.

Will add a note to the docs that the page size must be > the largest batch written.